src/lib/mongoBackend/mongoRegistrationsGet.cpp

```diff
@@ -0,0 +1,317 @@
+*
+* setStatus -
+*/
+static void setStatus(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+  regP->status = (r.hasField(REG_STATUS))? getStringFieldF(r, REG_STATUS): "";
+}
+
+
+
+/* ****************************************************************************
+*
+* setForwardingInformation -
+*/
+static void setForwardingInformation(ngsiv2::Registration* regP, const mongo::BSONObj& r)
+{
+  // No forwarding info until API V2 forwarding is implemented
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoRegistrationsGet - 
+*/
+void mongoRegistrationsGet
+(
+  std::vector<ngsiv2::Registration>*  regV,
+  const std::string&                  tenant,
+  const std::vector<std::string>&     servicePathV,
+  OrionError*                         oeP
+)
+{
+  bool         reqSemTaken = false;
+  std::string  err;
+  mongo::OID   oid;
+  StatusCode   sc;
+
+  reqSemTake(__FUNCTION__, "Mongo Get Registrations", SemReadOp, &reqSemTaken);
+
+  LM_T(LmtMongo, ("Mongo Get Registrations"));
+
+  std::auto_ptr<mongo::DBClientCursor>  cursor;
+  mongo::BSONObj                        q;
+
+  q = BSON(REG_SERVICE_PATH << fillQueryServicePath(servicePathV));
+
+  TIME_STAT_MONGO_READ_WAIT_START();
+  mongo::DBClientBase* connection = getMongoConnection();
+  if (!collectionQuery(connection, getRegistrationsCollectionName(tenant), q, &cursor, &err))
+  {
+    releaseMongoConnection(connection);
+    TIME_STAT_MONGO_READ_WAIT_STOP();
+    reqSemGive(__FUNCTION__, "Mongo Get Registrations", reqSemTaken);
+    oeP->fill(SccReceiverInternalError, err);
+    return;
+  }
+  TIME_STAT_MONGO_READ_WAIT_STOP();
+
+  /* Process query result */
+  while (moreSafe(cursor))
+  {
+    mongo::BSONObj        r;
+    ngsiv2::Registration  reg;
+
+    if (!nextSafeOrErrorF(cursor, &r, &err))
+    {
+      releaseMongoConnection(connection);
+      LM_E(("Runtime Error (exception in nextSafe(): %s - query: %s)", err.c_str(), q.toString().c_str()));
+      reqSemGive(__FUNCTION__, "Mongo Get Registrations", reqSemTaken);
+      oeP->fill(SccReceiverInternalError, std::string("exception in nextSafe(): ") + err.c_str());
+      return;
+    }
+    LM_T(LmtMongo, ("retrieved document: '%s'", r.toString().c_str()));
+
+    //
+    // Fill in the Registration with data retrieved from the data base
+    //
+    setRegistrationId(&reg, r);
+    setDescription(&reg, r);
+
+    if (setDataProvided(&reg, r, false) == false)
+    {
+      releaseMongoConnection(connection);
+      LM_W(("Bad Input (getting registrations with more than one CR is not yet implemented, see issue 3044)"));
+      reqSemGive(__FUNCTION__, "Mongo Get Registrations", reqSemTaken);
+      oeP->fill(SccReceiverInternalError, err);
+      return;
+    }
+
+    setExpires(&reg, r);
+    setStatus(&reg, r);
+    setForwardingInformation(&reg, r);
+
+    // FIXME PR: What about the Service Path of the Registration ... ?
```

Comment:
Nothing to do regarding service path.

Reply:
This is an old problems we've had since we introduced service paths.
If you issue a query with SP /# and you get a result, you don't know what service path each of the returned registrations belong to. I thought perhaps we'd fix this problem now. It's easy, just add the service path as a part of the payload and the user is fully informed ...

Reply:
NGSIv2 specification doesn't take into account that service path in the payload so the rigth way of implementing it without breaking the spec is not including it.

Thus, nothing to change here from my point of view (except removing the FIXME PR line).

Reply:
Moreover, note client don't need to get the service path in the response, as it is exactly the same he/she put in the request.

Remember that querying for subs or regs is not the same than queriying for entities, as explained at https://fiware-orion.readthedocs.io/en/master/user/service_path/index.html

> Fiware-ServicePath header is taken into account in `GET /v2/subscriptions` and `GET /v2/registrations` in order to narrow down the results to subscriptions/registrations that use *exactly* that service path as query.

(Deja vu feeling :)... we have already discussed this, see https://github.com/telefonicaid/fiware-orion/issues/3051)

Reply:
Lots to discuss here ...
If a reg exists with entity E and service path /a/# and a GET reg enters with service path /#, i.e. the user asks for a list of all registrations with **any** service path, should not that registration with service path /a/# be returned?
If we only return registrations with the **very exact** service path, not sure this service is very interesting ...
However, something is better than nothing, let's open an issue and postpone this question.
Implementing it with a "not very interesting" **exact** service path match.

Reply:
I think we already have an issue for this: telefonicaid#3051. See https://github.com/apinf/fiware-orion/pull/17#discussion_r166429735.

Reply:
The file is gone, and the FIXME with it